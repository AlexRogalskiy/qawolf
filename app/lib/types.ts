// GraphQL Types
export type AuthenticatedUser = {
  access_token: string;
  user: User;
};

export type DeploymentEnvironment = "all" | "preview" | "production";

export type Env = { [name: string]: string };

export type Environment = {
  id: string;
  name: string;
};

export type EnvironmentVariable = {
  created_at: string;
  environment_id: string;
  id: string;
  name: string;
  value: string;
};

export type Integration = {
  github_repo_name: string;
  id: string;
  slack_channel: string | null;
  team_name: string | null;
  type: "github" | "slack";
};

export type Invite = {
  email: string;
  id: string;
  team_id: string;
  wolf_name: string;
  wolf_number: number;
  wolf_variant: string;
};

export type RunStatus = "created" | "fail" | "pass";

export type Run = {
  code: string;
  completed_at: string | null;
  created_at: string;
  current_line: number | null;
  environment_id: string | null;
  id: string;
  logs_url: string | null;
  started_at: string | null;
  status: RunStatus;
  suite_id: string | null;
  test_id: string;
  video_url: string | null;
};

export type Runner = {
  api_key: string | null;
  ws_url: string | null;
};

export type ShortSuite = {
  created_at: string;
  environment_id: string | null;
  environment_variables: string | null;
  id: string;
  team_id: string;
  trigger_color: string | null;
  trigger_id: string | null;
  trigger_name: string | null;
};

export type Suite = ShortSuite & {
  runs: SuiteRun[];
};

export type SuiteRun = {
  completed_at: string | null;
  gif_url: string | null;
  id: string;
  is_test_deleted: boolean;
  started_at: string | null;
  status: RunStatus;
  test_id: string;
  test_name: string;
};

export type ShortTeam = {
  id: string;
  name: string;
};

export type Team = ShortTeam & {
  alert_integration_id: string | null;
  api_key: string;
  helpers: string;
  inbox: string;
  is_email_alert_enabled: boolean;
  is_enabled: boolean;
  next_trigger_id: string;
  renewed_at: string | null;
};

export type ShortTest = {
  id: string;
  name: string;
};

export type Test = ShortTest & {
  code: string;
  deleted_at: string | null;
  is_enabled: boolean;
  updated_at: string;
  version: number;
};

export type TestHistoryRun = {
  created_at: string;
  id: string;
  started_at: string;
  status: RunStatus;
};

export type TestSummaryRun = {
  created_at: string;
  gif_url: string | null;
  id: string;
  status: RunStatus;
};

export type TestSummary = {
  gif_url: string | null;
  last_runs: TestSummaryRun[];
  test_id: string;
};

export type Trigger = {
  color: string;
  created_at: string;
  deployment_branches: string | null;
  deployment_environment: DeploymentEnvironment | null;
  deployment_integration_id: string | null;
  environment_id: string | null;
  id: string;
  name: string;
  next_at: string | null;
  repeat_minutes: number | null;
};

export type User = {
  avatar_url: string | null;
  email: string;
  id: string;
  onboarded_at: string | null;
  teams: ShortTeam[];
  wolf_name: string;
  wolf_number: number;
  wolf_variant: string;
};

// Component Types
export type AuthMode = "logIn" | "signUp";

export type CreateCode = {
  code: string;
  test_id: string | null;
};

export type Modal =
  | "createTest"
  | "environments"
  | "deleteTests"
  | "teamSettings"
  | "triggers";

export type NavigationOption = "code" | "logs" | "helpers";

export type NavigationType = "dark" | "light";

export type SelectedTest = {
  id: string;
  name: string;
};

export type Side = "left" | "right";

export type TestTriggers = {
  test_id: string;
  trigger_ids: string[];
};

export type TriggerFields = {
  deployment_branches: string | null;
  deployment_environment: DeploymentEnvironment | null;
  deployment_integration_id: string | null;
  environment_id: string | null;
  name: string;
  repeat_minutes: number | null;
};

export type ValueProp = {
  detail: string;
  message: string;
};

export type Wolf = {
  name: string;
  number: number;
  variant: string;
};

// Runner Types
export type CodeUpdate = {
  code: string;
  generated?: boolean;
  test_id: string;
  version: number;
};

export type Log = {
  message: string;
  severity: "info" | "warning" | "error";
  timestamp: string;
};

export type RunOptions = {
  code: string;
  end_line?: number;
  env: Env;
  helpers?: string;
  restart: boolean;
  run_id?: string;
  start_line?: number;
  test_id?: string;
  version: number;
};

export type RunProgress = {
  code: string;
  completed_at: string | null;
  current_line: number | null;
  run_id?: string;
  start_line?: number;
  status: RunStatus;
};

// State Types
type SignUp = {
  inviteId?: string | null;
  redirectUri?: string | null;
};

type ModalState = {
  name: Modal | null;
  teamId?: string;
  testIds?: string[];
  tests?: SelectedTest[];
};

export type State = {
  dashboardUri: string | null;
  editorSidebarWidth: number;
  email: string | null;
  environmentId: string | null;
  modal: ModalState;
  run: RunOptions | null;
  signUp: SignUp;
  teamId: string | null;
  toast: {
    error?: boolean;
    expiresIn?: number;
    message: string;
  } | null;
};
