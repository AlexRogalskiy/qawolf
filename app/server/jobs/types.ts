export const JOB_TYPES = [
  "checkPending",
  "deleteOldEmails",
  "deleteRunners",
  "deployRunners",
  "orchestrateRunners",
  "orchestrateTriggers",
  "restartRunners",
] as const;

export type Job = typeof JOB_TYPES[number];
