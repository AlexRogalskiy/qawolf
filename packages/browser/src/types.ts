declare module "puppeteer/lib/USKeyboardLayout" {
  export interface KeyDefinition {
    code: string;
    key: string;
    keyCode?: number;
    location?: number;
    shiftKey?: string;
    shiftKeyCode?: number;
  }

  export interface KeyDefinitionsType {
    [key: string]: KeyDefinition;
  }

  const map: KeyDefinitionsType;
  export default map;
}
