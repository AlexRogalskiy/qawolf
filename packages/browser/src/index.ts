export { Page as PuppeteerPage } from "puppeteer";
export * from "./browser";
export * from "./find";
export * from "./retry";
export * from "./strokes";
