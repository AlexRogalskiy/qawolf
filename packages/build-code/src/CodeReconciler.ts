import { VirtualCode } from "./VirtualCode";

type ReconcileOptions = {
  actualCode: string;
  virtualCode: VirtualCode;
};

export class CodeReconciler {
  // the virtual code representation of the actual code
  private _virtualCode: VirtualCode;

  public hasUpdates(virtualCode: VirtualCode) {
    // const stepsToPatch = options.steps.slice(this._commitedStepIndex);
    // TODO....
    // if (!stepsToPatch.length && !options.removeHandle) {
    //   return;
    // }
    return false;
  }

  public reconcile({ actualCode, virtualCode }: ReconcileOptions): string {
    /**
     * Reconcile the actual code with the new virtual code.
     */
    let reconciledCode = actualCode;

    // update the last expression if it changed

    // patch new expressions

    return reconciledCode;
  }

  public update(virtualCode: VirtualCode) {
    this._virtualCode = virtualCode;
  }
}
