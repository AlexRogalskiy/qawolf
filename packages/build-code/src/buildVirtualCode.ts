import { Step } from "@qawolf/types";
import { AssertionExpression } from "./AssertionExpression";
import { ScriptExpression } from "./ScriptExpression";
import { StepExpression } from "./StepExpression";
import { VirtualCode } from "./VirtualCode";

export const buildVirtualCode = (steps: Step[], isTest: boolean = false) => {
  let previous: StepExpression;

  const expressions = steps.map(step => {
    const stepExpression = new StepExpression(step, previous);
    previous = stepExpression;

    if (isTest) {
      return new AssertionExpression(stepExpression);
    }

    return new ScriptExpression(stepExpression);
  });

  return new VirtualCode(expressions);
};
