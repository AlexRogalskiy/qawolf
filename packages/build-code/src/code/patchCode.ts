import { buildStepsCode } from "../build";
import { getIndentation, indent, removeLinesIncluding } from "../format";

export const PATCH_SYMBOL = "// 🐺 CREATE CODE HERE";

export const buildPatch = () => {
  let patch = buildStepsCode({
    startIndex,
    steps,
    isTest
  });

  // include the symbol to replace later
  patch += PATCH_SYMBOL;

  return patch;
};

export const canPatch = (code: string) => {
  return code.includes(PATCH_SYMBOL);
};

export const indentPatch = (code: string, patch: string) => {
  /**
   * Match the indentation of the symbol.
   */
  const numSpaces = getIndentation(code, PATCH_SYMBOL);

  return indent(patch, numSpaces, 1);
};

export const patchCode = ({ code, startIndex, steps, isTest }) => {
  let patch = buildPatch();

  patch = indentPatch(code, patch);

  const patchedCode = code.replace(PATCH_SYMBOL, patch);

  if (removeSymbol) {
    code = removeLinesIncluding(code, PATCH_SYMBOL);
  }

  return patchedCode;
};
