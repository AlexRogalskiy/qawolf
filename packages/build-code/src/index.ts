export * from "./buildInitialCode";
export { stepToSelector } from "./buildSelector";
export * from "./buildVirtualCode";
export * from "./CodeReconciler";
