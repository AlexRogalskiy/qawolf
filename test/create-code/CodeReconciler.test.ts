import { loadFixtures } from '../loadFixtures';
import { buildVirtualCode } from '../../src/build-code/buildVirtualCode';
import { buildSteps } from '../../src/build-workflow/buildSteps';
import { CodeReconciler } from '../../src/create-code/CodeReconciler';
import { PATCH_HANDLE } from '../../src/create-code/patchCode';
import { Step } from '../../src/types';

describe('CodeReconciler.reconcile', () => {
  let steps: Step[];

  beforeAll(async () => {
    const events = (await loadFixtures('login')).events;
    steps = buildSteps(events);
  });

  it('ignores new expressions when patch handle is missing', () => {
    const reconciler = new CodeReconciler();

    const newCode = buildVirtualCode([steps[0]]);
    expect(reconciler.hasChanges(newCode)).toBeTruthy();

    const reconciled = reconciler.reconcile({
      actualCode: 'original code',
      virtualCode: newCode,
    });
    expect(reconciled).toEqual('original code');
  });

  it('inserts new expressions', () => {
    const reconciler = new CodeReconciler();

    const newCode = buildVirtualCode([steps[0]]);
    expect(reconciler.hasChanges(newCode)).toBeTruthy();

    const reconciled = reconciler.reconcile({
      actualCode: PATCH_HANDLE,
      virtualCode: newCode,
    });
    expect(reconciled).toMatchSnapshot();
  });

  it('updates the last expression', () => {
    const reconciler = new CodeReconciler();

    const originalCode = buildVirtualCode([steps[0]]);

    const actualCode = reconciler.reconcile({
      actualCode: PATCH_HANDLE,
      virtualCode: originalCode,
    });

    reconciler.update(originalCode);

    const newCode = buildVirtualCode([steps[1]]);
    expect(reconciler.hasChanges(newCode)).toBeTruthy();

    const reconciled = reconciler.reconcile({
      actualCode,
      virtualCode: newCode,
    });
    expect(reconciled).toMatchSnapshot();
  });

  it('updates the last expression after a step was removed (ex: paste)', () => {
    const reconciler = new CodeReconciler();

    const originalCode = buildVirtualCode([steps[0]]);

    const actualCode = reconciler.reconcile({
      actualCode: PATCH_HANDLE,
      virtualCode: originalCode,
    });

    reconciler.update(originalCode);

    const removedStepCode = buildVirtualCode([]);

    // we don't want it to do anything until a new step arrives
    expect(reconciler.hasChanges(removedStepCode)).toBeFalsy();
    let reconciled = reconciler.reconcile({
      actualCode,
      virtualCode: removedStepCode,
    });
    expect(reconciled).toEqual(actualCode);

    reconciler.update(removedStepCode);

    const stepAddedCode = buildVirtualCode([steps[1]]);
    expect(reconciler.hasChanges(stepAddedCode)).toBeTruthy();

    reconciled = reconciler.reconcile({
      actualCode,
      virtualCode: stepAddedCode,
    });
    expect(reconciled).toMatchSnapshot();
  });
});
