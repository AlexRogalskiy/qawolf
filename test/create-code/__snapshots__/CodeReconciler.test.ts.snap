// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodeReconciler.reconcile inserts new expressions 1`] = `
"await page.click(\\"text=Log in\\");
// 🐺 create code here"
`;

exports[`CodeReconciler.reconcile updates the last expression 1`] = `
"await page.click(\\"#username\\");
// 🐺 create code here"
`;

exports[`CodeReconciler.reconcile updates the last expression after a step was removed (ex: paste) 1`] = `
"await page.click(\\"#username\\");
// 🐺 create code here"
`;
